src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -451,14 +459,21 @@ static void init_keepalive_pings_if_enabled(grpc_chttp2_transport* t) {
   if (t->keepalive_time != grpc_core::Duration::Infinity()) {
     t->keepalive_state = GRPC_CHTTP2_KEEPALIVE_STATE_WAITING;
     GRPC_CHTTP2_REF_TRANSPORT(t, "init keepalive ping");
-    GRPC_CLOSURE_INIT(&t->init_keepalive_ping_locked, init_keepalive_ping, t,
-                      grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&t->keepalive_ping_timer,
-                    grpc_core::Timestamp::Now() + t->keepalive_time,
-                    &t->init_keepalive_ping_locked);
+    // The timer can fire on an EventEngine thread before RunAfter returns,
+    // causing issue with overriding the handle. Thus use a mutex here.
+    // Hopefully this only happens once during init.
+    std::shared_ptr<absl::Mutex> mu = std::make_shared<absl::Mutex>();
+    absl::MutexLock lock(&*mu);
+    t->keepalive_ping_timer_handle =
+        t->event_engine->RunAfter(t->keepalive_time, [t, mu] {
+          grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+          grpc_core::ExecCtx exec_ctx;
+          absl::MutexLock lock(&*mu);
+          init_keepalive_ping(t);
```

Comment:
Why not just re-enter the combiner here, eschewing the mutex?

Reply:
yep, that's the other approach I was considering... `init_keepalive_ping` does re-enter the combiner, we need to run `init_keepalive_pings_if_enabled` in the combiner, so that `init_keepalive_ping_locked` is guaranteed to run after `init_keepalive_pings_if_enabled` had returned.

I decide to make the change since your prompt has changed my mind. :) PTAL. And also I don't like introducing mutex to chttp2_transport and adding performance overhead potentially.

This approach might slightly delay the run of the keepalive_ping_timer though.

Reply:
I think that delay is probably ok, but I like that we keep a consistent locking story in this code (we want to change that story, and having uniformity ahead of time is a huge boon)

Reply:
Gotcha, thanks Craig!